src/mon/OSDMonitor.cc

```diff
@@ -1663,6 +1665,12 @@ bool OSDMonitor::check_failure(utime_t now, int target_osd, failure_info_t& fi)
     for (map<int,failure_reporter_t>::iterator p = fi.reporters.begin();
 	 p != fi.reporters.end();
 	 ++p) {
+      //get subtree reporter
+      map<string,string> reporter_loc = osdmap.crush->get_full_location(p->first);
+      int osd_subtree_reporter  = osdmap.crush->get_item_id(reporter_loc[reporter_subtree_level]);
+      assert(osd_subtree_reporter != 0);
+      subtree_reporters.insert(osd_subtree_reporter);
```

Comment:
what i am doing here is a bit trick, actually it's on check_failure so the cancel/count is done in prepare_failure by OSD level, and in check_failure we just consolidate them into subtree level. I think this is simpler although waste a bit calculation, but usually there will not be too much report and run every tick (5s).